### ISSUE TITLE
Efficient forward pass in nn.Linear

### ISSUE THREAD

**Issue by yunjey** [pytorch/pytorch#2159]
Hi, 

Maybe we can improve the operation speed in `nn.Linear`.

`nn.Linear` creates a weight matrix of shape (out_features, in_features) as you can see [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/linear.py#L39).
Then, as you can see in [F.linear](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L550-L558), input matrix is multiplied by the transpose of the weight matrix.

I think the transpose operation is unnecessary. What if we create a weight matrix of shape (in_features, out_features)? Then, the input matrix can be multiplied by the weight matrix (we don't need to calculate transpose of the matrix).

Does this sound right? If i'm right, can i do pull request?

Thanks

**Comment by soumith** [comment]
the transpose is completely free, so i dont think the PR is necessary. On the backend side, BLAS natively supports transposed matrices, so this is not a perf gain or loss.

**Comment by ngimel** [comment]
Typically, nn kernels are slightly faster than nt, so it might be a slight performance gain, but then backward multiplies might become less efficient than they are now, because now transposes in backward will be different.  FWIW, cudnn rnn, when doing a longer time sequence, starts by transposing weight matrix, and in most cases it's a net performance gain, but it's doing this for forward only.

**Comment by carlosgmartin** [comment]
@soumith But why? It's a bit annoying to have to swap dimensions and deal with transposition when doing custom weight initialization. Is there a rationale for the current approach?

**Comment by soumith** [comment]
@carlosgmartin it's historical weight layout, changing it is backward-incompatible. Unless there is some BIG benefit in terms of speed or convenience, we wont break userland.

**Comment by kushalj001** [comment]
Is this behavior only restricted for `nn.Linear` layers or is it implemented in all `nn` modules. I specifically want to know if the internal weight matrices are transposed for an `RNN` layer. I can see that the `weight_ih` that is input to hidden matrix is transposed while storing but I cannot be sure about the `weight_hh` since it's a square matrix. I need to know since I am updating weight manually for each connection and transposed matrices might imply that I am updating the wrong connections. Basically I want to know "which neuron led the neruon in subsequent layer to fire".
